sql/sql_connect.cc

```diff
@@ -84,6 +84,20 @@ static int increment_connection_count(THD* thd, bool use_lock);
 
 static HASH hash_user_connections;
 
+int get_user_con_by_name(const char *user) {
```

Comment:
Please re-check the hash function for hash_user_connections whether it's really based on user@host, and not user alone: it is initialised as

```
    my_hash_init(&hash_user_connections,system_charset_info,max_connections,
                 0,0, (my_hash_get_key) get_key_conn,
                 (my_hash_free_key) free_user, 0);
```

and get_key_conn is

```
extern "C" uchar *get_key_conn(user_conn *buff, size_t *length,
			      my_bool not_used __attribute__((unused)))
{
  *length= buff->len;
  return (uchar*) buff->user;
}
```

which suggests that is hashes on the username along.

If not, then I'm afraid we cannot have O(n) here, this will hurt with high concurrency / high userstat query ratio.